File: python/problem1504-1.py
class Solution:
    def numSubmat(self, mat: List[List[int]]) -> int:
        height = mat#copy.deepcopy(mat)
        row, col = len(mat), len(mat[0])
        for i in range(1, row):
            for j in range(col):
                if height[i][j] == 1:
                    height[i][j] = height[i - 1][j] + 1
        result = 0
        for i in range(row):
            for j in range(col):
                if height[i][j] == 0:
                    continue
                left = j
                minHeight = height[i][j]
                while left >= 0 and height[i][left] > 0:
                    result += minHeight
                    left -= 1
                    minHeight = min(minHeight, height[i][left])
        return result

File: Java/problem1504-1.java
class Solution {
    public int numSubmat(int[][] mat) {
        int row = mat.length, col = mat[0].length;
        for (int i = 1; i < row; ++i) {
            for (int j = 0; j < col; ++j) {
                if (mat[i][j] == 0) {
                    continue;
                }
                mat[i][j] = mat[i - 1][j] + 1;
            }
        }
        int result = 0;
        for (int i = 0; i < row; ++i) {
            for (int j = 0; j < col; ++j) {
                if (mat[i][j] == 0) {
                    continue;
                }
                int minHeight = mat[i][j];
                int left = j;
                while (left >= 0 && mat[i][left] > 0) {
                    minHeight = Math.min(minHeight, mat[i][left--]);
                    result += minHeight;
                }
            }
        }
        return result;
    }
}
